src/Moq/InvocationShape.cs

```diff
@@ -109,5 +110,71 @@ private bool IsOverride(MethodInfo invocationMethod)
 
 			return true;
 		}
+
+		public bool Equals(InvocationShape other)
+		{
+			if (this.Method != other.Method)
+			{
+				return false;
+			}
+
+			if (this.Arguments.Count != other.Arguments.Count)
+			{
+				return false;
+			}
+
+			if (this.partiallyEvaluatedArguments == null)
+			{
+				this.partiallyEvaluatedArguments = PartiallyEvaluateArguments(this.Arguments);
```

Comment:
It might be good to someday think about whether this needs additional precautions for thread-safety. For now I'll pretend to be naïve and think that this operation would yield the same result even when called several times by several threads, so it's fine to just let the last evaluation win.